client/state/utils.js

```diff
@@ -474,6 +474,40 @@ function serializeState( reducers, state, action ) {
 	);
 }
 
+function applyStoredState( reducers, state, action ) {
+	let hasChanged = false;
+	const nextState = mapValues( reducers, ( reducer, key ) => {
+		// Replace the value for the key we want to init with action.storedState.
+		if ( reducer.storageKey === action.storageKey ) {
+			hasChanged = true;
+			return action.storedState;
+		}
+
+		// Descend into nested state levels, possibly the storageKey will be found there?
+		const prevStateForKey = state[ key ];
+		const nextStateForKey = reducer( prevStateForKey, action );
+		hasChanged = hasChanged || nextStateForKey !== prevStateForKey;
+		return nextStateForKey;
+	} );
+
+	// return identical state if the stored state didn't get applied in this reducer
+	return hasChanged ? nextState : state;
+}
+
+function getStorageKeys( reducers ) {
+	return function*() {
```

Comment:
Is there a reason why we use a generator function and not e.g. an async function?

Reply:
`getStorageKeys` traverses the reducer tree and gathers the `storageKey` attributes into a linear array. It's synchronous, nothing async at all.

Generator seemed to me like a very nice way to traverse the tree and concatenate values and lists of values into one big list.

Reply:
I can imagine three ways of doing the tree traversal:

Generator:
```js
function*() {
  for ( const reducer of Object.values( reducers ) ) {
    if ( reducer.storageKey ) yield reducer.storageKey;
    if ( reducer.getStorageKeys ) yield* reducer.getStorageKeys();
  }
}
```

Functional version that uses `flatMap`:
```js
function() {
  return flatMap( reducers, reducer => flatMap( [
    reducer.storageKey ? [ reducer.storageKey ] : [],
    reducer.getStorageKeys ? reducer.getStorageKeys() : []
  ] ) );
}
```

Build an output array using `push` and `concat`:
```js
function() {
  let rv = [];
  for ( const reducer of Object.values( reducers ) ) {
    if ( reducer.storageKey ) rv.push( reducer.storageKey );
    if ( reducer.getStorageKeys ) rv = rv.concat( reducer.getStorageKeys() );
  }
  return rv;
}
```

I found the generator version most likeable: it's easy to read (given the reader is fluent in generators of course) and it also shouldn't allocate any ephemeral arrays (not sure if it matters performance-wise though -- maybe the Babel-transpiled generator has its own overhead that's not immediately visible).